### ISSUE TITLE
Debugger breakpoints using incorrect URL for a remote debugging session

### ISSUE THREAD

**Issue by treffynnon** [ChromeDevTools/devtools-protocol#103]
# Symptoms
This is related to the closed ticket #99, which describes the symptoms of this very problem.

You set a breakpoint, but the process isn't paused when that breakpoint is encountered and carries merrily on.

# Environment
This is a remote debugging session between a Windows client (running Chrome devtools) and a remote Linux server (running node with `--inspect=0.0.0.0`).

# What's wrong
By watching the websocket communication with Wireshark it is apparent that the Chrome Dev tools are using an incorrect filepath/URL when calling `Debugger.setBreakpointByUrl`.

### Request
```json
{
  "id": 17,
  "method": "Debugger.setBreakpointByUrl",
  "params": {
    "lineNumber": 67,
    "url": "EFS\\DEV-efsbastion\\username\\application\\projects\\service\\module\\index.js",
    "columnNumber": 18,
    "condition": ""
  }
}
```

### Response
```json
{
  "id": 17,
  "result": {
    "breakpointId": "EFS\\DEV-efsbastion\\username\\application\\projects\\service\\module\\index.js:67:18",
    "locations": []
  }
}
```

**Note** the empty locations array. I am no node debugger expert, but it feels like this locations array should be checked to ensure it is not empty after setting a new breakpoint marker before then showing the marker as active in the UI.

### Request
```json
{
  "id": 17,
  "method": "Debugger.setBreakpointByUrl",
  "params": {
    "lineNumber": 67,
    "url": "/EFS/DEV-efsbastion/username/application/projects/service/module/index.js",
    "columnNumber": 18,
    "condition": ""
  }
}
```

### Response
```json
{
  "id": 17,
  "result": {
    "breakpointId": "/EFS/DEV-efsbastion/username/application/projects/service/module/index.js:67:18",
    "locations": [{
      "scriptId":"104",
      "lineNumber":67,
      "columnNumber":18
    }]
  }
}
```

**Note** the Linux path separators instead of Windows and the initial root slash on the path.

# But isn't this a node bug?

No, it is not from what I can tell because of the following two discoveries.

## PHPStorm
PHPStorm/WebStorm debugger continues to work just fine using it's URL regexes

### Request
```json
{
  "id": 17,
  "method": "Debugger.setBreakpointByUrl",
  "params": {
    "lineNumber": 67,
    "urlRegex": "[/\\\\][iI][nN][dD][eE][xX]\\.[jJ][sS]([;?#!].*)?$"
  }
}
```

### Response
```json
{
  "id": 17,
  "result": {
    "breakpointId": "/[/\\\\][iI][nN][dD][eE][xX]\\.[jJ][sS]([;?#!].*)?$/:67:18",
    "locations": [{
      "scriptId":"104",
      "lineNumber":67,
      "columnNumber":18
    }]
  }
}
```
## Manually recreating the request successfully
If I manually set the breakpoint myself I see a success response back over the websocket. For example consider the following code run from the Chrome console:

```javascript
let id, key, socket

id = 1000
key = '<debugger-key-goes-here>'

const connect = userSuppliedKey => {
	socket = new WebSocket(`ws://<ip-address-goes-here>:9229/${userSuppliedKey || key}`)
	socket.addEventListener('message', e => {
		const d = JSON.parse(e.data)
		if (d) {
			if (d.method === 'Debugger.scriptParsed') return console.log('scriptParsed', JSON.stringify({ scriptId: d.params.scriptId, url: d.params.url }))
			if (d.method === 'Debugger.breakpointResolved') return console.log('breakpointResolved', JSON.stringify(Object.assign({ breakpointId: d.params.breakpointId }, d.params.location)))
		}
		console.info(`${e.data}`)
	})
}
const send = (method, x) => socket.send(JSON.stringify({ id: ++id, method: `Debugger.${method}`, params: x }))
const _getPossibleBps = (start, end) => send('getPossibleBreakpoints', { start, end })
const getPossibleBps = (scriptId, start, end) => _getPossibleBps({ scriptId, lineNumber: start }, { scriptId, lineNumber: end })
const setBp = (lineNumber, url) => send('setBreakpointByUrl', { lineNumber, url, })
const removeBp = (lineNumber, url, columnNumber) => send('removeBreakpoint', { breakpointId: `${url}:${lineNumber}:${columnNumber || 0}` })
const enable = () => send('enable')
```

Which gives us this when run in the Chrome console

```javascript
connect('<debugger-key-goes-here>')
// '<debugger-key-goes-here>'

enable()
// ...
// scriptParsed {"scriptId":"104","url":"/EFS/DEV-efsbastion/username/application/projects/service/module/index.js"}
// ...

getPossibleBps('104', 9, 15)
// {"id":1016,"result":{"locations":[{"scriptId":"104","lineNumber":9,"columnNumber":15},{"scriptId":"104","lineNumber":11,"columnNumber":15},{"scriptId":"104","lineNumber":12,"columnNumber":16},{"scriptId":"104","lineNumber":14,"columnNumber":4},{"scriptId":"104","lineNumber":15,"columnNumber":0},{"scriptId":"104","lineNumber":17,"columnNumber":4},{"scriptId":"104","lineNumber":18,"columnNumber":14}]}}

setBp(11, '/EFS/DEV-efsbastion/username/application/projects/service/module/index.js')
// {"id":1024,"result":{"breakpointId":"/EFS/DEV-efsbastion/username/application/projects/service/module/index.js:11:0","locations":[{"scriptId":"104","lineNumber":11,"columnNumber":15}]}}

removeBp(11, '/EFS/DEV-efsbastion/username/application/projects/service/module/index.js')
// {"id":1019,"result":{}}
```

**Comment by treffynnon** [comment]
Also reference on https://bugs.chromium.org/p/chromium/issues/detail?id=841697

**Comment by ak239** [comment]
Thanks for great report!

Inside V8 on inspector backend side we try to match breakpoint url and script url. In this case we send incorrect url from frontend side and when we receive new parsed script on backend side we do not set breakpoint properly.

I will do following two points next week:
- check why we send incorrect url from frontend side,
- consider some kind of breakpoint url normalization on backend side to make our protocol a little more flexible.

I will update this issue as soon as I get some results.

(PhpStorm actually uses great regexp)

**Comment by mathieulj** [comment]
Just to corroborate that this is due to a change on the chrome side and provide a workaround. We experienced this recently running windows workstations with a linux development VM. We had 3 versions of chrome(64, 66, 68) connecting one at a time to a debug session and only 64 was working normally. This was true for node 6, 8 and 10.

As a workaround for those who need to debug, we found that using the id and manually visiting the link `chrome-devtools://devtools/bundled/inspector.html?experiments=true&v8only=true&ws=${ip}:${port}/${id}` works as expected even in the newer chrome versions. If, for example, you get the message `Debugger listening on ws://0.0.0.0:3020/4dacc23a-f296-4b6d-bb47-18af21dfef67` during inspect start, the URL would be `chrome-devtools://devtools/bundled/inspector.html?experiments=true&v8only=true&ws=${ip}:3020/4dacc23a-f296-4b6d-bb47-18af21dfef67` where ip is the ip of the remote machine. We also found that the plugin nodejs-v8-inspector-manag(https://chrome.google.com/webstore/detail/nodejs-v8-inspector-manag/gnhhdgbaldcilmgcpfddgdbkhjohddkj) seems to still work.

**Comment by brucejo75** [comment]
Also see: https://bugs.chromium.org/p/chromium/issues/detail?id=844070
& https://github.com/meteor/meteor/issues/9875

**Comment by tonymet** [comment]
I'm experiencing the lack-of-breakpoints with Windows 10 + WSL (aka Ubuntu Bash for Windows) + Node 8.11.3

@mathieulj  's workaround by manually building the `chrome-devtools://` URL works for me.

Thanks for the help guys this was driving me bonkers for a while :D

Node v8.11.3
Windows 10 PRo 1803
Version 67.0.3396.87 (Official Build) (64-bit)

**Comment by tonymet** [comment]
Also mentioned in the chromium bug, Chrome Extensions "Inspector Manager" is a simpler workaround for now

https://chrome.google.com/webstore/detail/nodejs-v8-inspector-manag/gnhhdgbaldcilmgcpfddgdbkhjohddkj

**Comment by treffynnon** [comment]
@ak239 is there any progress being made with this issue?

Unfortunately, [Web/PHP]Storm has it's own issues with simply choosing the first file it sees with the name the breakpoint is set against. In the case of files named `index.js` this is an issue as so many modules use this filename that the breakpoint often ends up being set against the wrong file.

**Comment by ak239** [comment]
I landed following [CL](https://chromium-review.googlesource.com/c/chromium/src/+/1130936) recently. It should fix provisional breakpoints in Node.js targets. At least it helps [ndb](https://github.com/GoogleChromeLabs/ndb).
I have another CL to land that fixes breakpoints in experimental Node.js modules: [CL](https://chromium-review.googlesource.com/c/chromium/src/+/1152906). I am going to land it soon.
It would be nice if someone could verify that issue is fixed in latest Canary.

**Comment by treffynnon** [comment]
I downloaded the canary with version: Version 70.0.3508.0 (Official Build) canary (64-bit) and it appears to have the same problem if that helps.

**Comment by mathieulj** [comment]
It seems to have gotten worse with the release of chrome 69. The workarounds I posted above no longer work with chrome 69.0.3497.81(stable) and with chrome 71.0.3548.0 (canary). Workstations still running 68.0.3440.106 do work with both workarounds.

**Comment by mathieulj** [comment]
For those with a Webstorm licence, version 2018.2.3 seems to break correctly on breakpoints.

**Comment by treffynnon** [comment]
For those without a WebStorm licence you can use the [Early Access Programme (EAP) builds](https://www.jetbrains.com/webstorm/eap/) for free with the following notice:

>This is an early access version of the product. You expressly acknowledge that this version of the product may not be reliable, may not work as intended and may contain errors. Any use of the EAP product is at your own risk.
>
> -- JetBrains

**Comment by ak239** [comment]
fileURL patches [were merged](https://github.com/nodejs/node/pull/22918) to Node 10. So I expect this issue to be fixed in next Node 10 release, I will update this bug when release happens.

**Comment by lantern77** [comment]
Will this be updated for node 8 at well? As I just recently encountered the bug after Chrome updated XD

**Comment by june07** [comment]
The most recent version of NiM actually does work as a fix to this.  Different DevTools versions can be selected along with a custom version as the video shows.  https://youtu.be/hjf6esnqOJQ

So in essence when using NiM it doesn't matter which version of Node or Chrome you use.

**Comment by sdd** [comment]
@june07 ?I'm using NiM 1.1.0, Node 10.12 (remote, linux), Chrome (69, windows) and I still get this problem

**Comment by june07** [comment]
You need to change the DevTools version under NiM settings...

**Comment by sdd** [comment]
@june07 thanks for the quick response.

I'v tried changing NiM settings but it is not clear what to use for the value of custom devtools.

I have tried a few values but none work, and now if I try to turn off "use alternate devtools", it does not open an inspector at all.

**Comment by june07** [comment]
I've just tried NiM 1.1.0 with Node 10.12 and Version 69.0.3497.100
(Official Build) (64-bit) and can confirm that:

https://chrome-devtools-frontend.appspot.com/serve_file/@a10b9cedb40738cb152f8148ddab4891df876959/inspector.html

works.  Which is the appspot option.





[image: Mailtrack]
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality6&>
Sender
notified by
Mailtrack
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality6&>
10/12/18,
12:31:15 PM

**Comment by smeijer** [comment]
Okay, let me rephrase that. WebStorm does kind of work. Some breakpoints can be hit, other ones not at all. And others don't seem to be hit, but my frontend does block. So node is pausing and waiting for the debugger, but I don't have control over it.

So I basically have 3 types of breakpoints.

1) The ones that function as expected
2) Those that cannot be hit
3) Those that are hit, but not given control over?

**Comment by brucejo75** [comment]
@ak239, Should I assume that Node 8 will not be supported?

I've just tried NiM 1.1.0 with Node 8.11.3 and Version 69.0.3497.100
(Official Build) (64-bit) and it breakpoints are not hit.

**Comment by ak239** [comment]
I definitely would like to get if fixed, but I need your help.
@brucejo75, could you provide more details about your setup, I assume that you use Windows and can reproduce this bug with any node script with node 8? (I tried on mac and it works).
Could you confirm that with latest node 10.12 it works to you? And that it does not work with dedicated frontend from latest [Google Chrome Canary](https://www.google.com/intl/en_ca/chrome/canary/)?

**Comment by ak239** [comment]
Could we focus on fixing DevTools and DevTools protocol in this issue please? A lot of workarounds were already mentioned and I believe that now we can focus on finding solution for root issue.

**Comment by brucejo75** [comment]
@june07, 

I had the devtools set correctly but failed to turn them on, D'oh.

When I use it now, I notice that the debug console uses v66 of chrome and all works as expected.

**Comment by june07** [comment]
> @june07,
> 
> I had the devtools set correctly but failed to turn them on, D'oh.
> 
> When I use it now, I notice that the debug console uses v66 of chrome and all works as expected.

Actually updated NiM (1.2.0) to use the most up to date version of DevTools (Chrome 71) as well as adding info on how to better use the custom setting https://june07.com/blog/nim-custom-devtools-url.

@ak239 This will be my last post to this thread.  Cheers.

**Comment by brucejo75** [comment]
Oops! spoke too soon.  devTools fails to stop on a breakpoint in a required module.
@ak239, fails with Canary :-(
@june07, also fails with https://chrome-devtools-frontend.appspot.com/serve_file/@a10b9cedb40738cb152f8148ddab4891df876959/inspector.html

### Environment
**Windows 10**: Version 1803 (OS Build 17134.345)
**node**: 8.11.3 (using WSL)
**chrome**: 72.0.3580.0 (Official Build) canary (64-bit)

**Note**: I am running WSL on windows, which is the Ubuntu linux subsystem.

### node script index.js:
```javascript
1 let foo = require('./foo');
2 console.log('hello World');
3 foo();
```
### node script foo.js
```javascript
1 function foo() {
2   console.log('Hit 1?');
3   debugger;
4   console.log('Hit 2?');
5 }
6
7 module.exports = foo;
```
### Repro steps
1. `node --inspect-brk index.js`
2. in chrome browser go to `chrome://inspect`, click on `Open dedicated DevTools for Node`
3. Set breakpoint on `foo.js` Line 2. 
4. in dedicated devTools, resume script execution (e.g. hit `F8`)

**Result**: devTools fails to stop on the breakpoint, stops on `foo.js` line 3.

### Workaround
1. `node --inspect-brk index.js`
2. in chrome browser go to `chrome://inspect`, click on `Open dedicated DevTools for Node`
3. set breakpoint on `index.js` line 2
4. set breakpoint on `foo.js` line 2.
5. in dedicated devTools, resume script execution (e.g. hit `F8`)

**Result**: devTools will now stop on both breakpoints.

**Comment by cronvel** [comment]
On Chrome/Chromium v69, UI breakpoints (not debugger statement) doesn't work anymore since node v10. 
It was working on node v8 (and I confirm it's still working when I downgrade). I believe it's a bug on the devTools side nonetheless. Also in Node v10, the devTools doesn't skip Node.js source code anymore, which is pretty annoying...

**Comment by brucejo75** [comment]
I verified that Chrome release version: `Version 71.0.3578.80 (Official Build) (64-bit)` now has this repaired.  Maybe should be closed?

**Comment by daniele-orlando** [comment]
I confirm that Chrome `71.0.3578.98` works fine.

**Comment by louderspace** [comment]
I'm using v8 8.0.426 and am still seeing this issue with the latest Chrome and VS Code.  Installing NiM did fix the problem on the mac, but not on windows  (though, I'm not convinced NiM is correctly installed on windows because I'm not seeing the syntax coloring).  I did verify that all of my paths are correctly specified within the v8::ScriptOrigin.


Here's an example:

Ticks = 23110      <17356> <V8> <5> messageFromDebugger : /1, payload : {"id":17,"method":"Debugger.setBreakpointByUrl","params":{"urlRegex":"[Dd]:\\\\dev\\\\apd\\\\<...redacted...>\\.js|file:\\/\\/\\/[Dd]:\\/dev\\/apd\\/<...redacted...>\\.js","lineNumber":430,"columnNumber":13}}

**Comment by june07** [comment]
What version of DevTools, NiM, and Node.js are you running on Windows?

**Comment by louderspace** [comment]
Node.js v12.16.3
NiM 2.3.4
Google Chrome | 81.0.4044.138
VS Code 1.45.1

Note, I've embedded v8 into a desktop application directly.

Updating Node.js yesterday helped on windows, but I'm still seeing the problem within VSCode.

**Comment by june07** [comment]
Well regarding

> (though, I'm not convinced NiM is correctly installed on windows because I'm not seeing the syntax coloring)

I can confirm running with the same versions as you that syntax coloring is working in both Chrome 81.0.4044.138 and Edge Version 83.0.478.28 (using both currently).

If that is still an issue you may want to double check your actual DevTools version, as in the case of NiM it's possible you have it set several different ways.:
![image](https://user-images.githubusercontent.com/11353590/82486114-43674c00-9a91-11ea-9990-27991059d300.png)

**Comment by louderspace** [comment]
I'm getting a new machine today and will try setting it up again from scratch.  Here's what it currently looks like.

![image](https://user-images.githubusercontent.com/62191204/82488638-161c9d00-9a95-11ea-9405-74a35cfd4fa3.png)

**Comment by ak239** [comment]
@louderspace If you are embedding V8 directly, your ScriptOrigin might contain properly formed file URLs like file://... - in this case, everything should work out of the box.

The solution that was implemented on the DevTools side only works for Node.js. So how do you start DevTools to debug your embedded V8?

**Comment by louderspace** [comment]
I am setting the script origins:

Added tracing shows "SetScriptPath = D:\dev\p4\<redacted>\nml\nml_include.js

Should this be in the form file://dev/... ?

**Comment by louderspace** [comment]
I tried various flavors and haven't had any success:

CompileAndRun script : file:///D:/dev/p4<redacted>/nml/nml_include.js

I do have NiM working on windows now and running Chrome 8.3
